drivers/pms5003/pms5003.c

```diff
@@ -0,0 +1,468 @@
+#define PMS5003_PARSER_RESP_CRC_IDX      3U
+#define PMS5003_PARSER_PM_CRC_IDX        15U
+
+/* the response timeout is based on 2.3 seconds being the maximum time
+   between measurements */
+#define PMS5003_RESP_TIMEOUT        (3 * US_PER_SEC)
+/* this is not documented in the datasheet but after a wakeup the device
+   is not fully available before 3 seconds */
+#define PMS5003_WAKEUP_TIME         (3 * US_PER_SEC)
+/* this is not documented in the datasheet but commands can't be sent out
+   within a short time of each other */
+#define PMS5003_CMD_BUSY_TIME       (50 * US_PER_MS)
+
+/**
+ * @name      UART commands for pms5003 driver
+ * @{
+ */
+typedef enum {
+    PMS5003_CMD_PASSIVE_READ    = 0xe2, /**< Perform a passive read */
+    PMS5003_CMD_SET_MODE        = 0xe1, /**< Set operating mode */
+    PMS5003_CMD_SET_SLEEP       = 0xe4, /**< Set sleep state*/
+} pms5003_cmd_t;
+
+/**
+ * @brief   Response to PMS5003_CMD_SET_MODE active
+ */
+#define PMS5003_CMD_ACTIVE_RESPONSE         0x01E1
+/**
+ * @brief   Response to PMS5003_CMD_SET_MODE passive
+ */
+#define PMS5003_CMD_PASSIVE_RESPONSE        0x00E1
+/**
+ * @brief   Response to PMS5003_CMD_SET_SLEEP sleep (wakeup does nto issue
+ *          a response)
+ */
+#define PMS5003_CMD_SLEEP_RESPONSE          0x00E4
+
+/**
+ * @brief   Frame parser state
+ *
+ * The different states match the index of the data in the received frame.
+ */
+typedef enum {
+    PMS5003_STATE_START_BYTE_1  = 0,                            /**< Waiting for first start byte */
+    PMS5003_STATE_START_BYTE_2,                                 /**< Waiting for second start byte */
+    PMS5003_STATE_FRAME_LENGTH_H,                               /**< Higher byte of frame length */
+    PMS5003_STATE_FRAME_LENGTH_L,                               /**< Lower lower byte of frame length*/
+    PMS5003_STATE_RESP_CRC_L    = PMS5003_RESP_FRAME_LEN - 1,   /**< Last byte of CMD response, lower crc byte */
+    PMS5003_STATE_PM_CRC_L      = PMS5003_PM_FRAME_LEN - 1,     /**< Last byte of PM data, lower crc byte */
+} pms5003_parser_state_t;
+
+static int _wait_for_resp(pms5003_t *dev)
+{
+    /* acquire lock that will be freed after a uart cmd response is
+       received */
+    mutex_lock(&dev->internal.timeout);
+    DEBUG_PUTS("[pms5003]: response wait");
+    /* try to acquire lock with a timeout in case no response is received */
+    int ret = ztimer_mutex_lock_timeout(ZTIMER_USEC, &dev->internal.timeout,
+                                        PMS5003_RESP_TIMEOUT);
+    if (ret) {
+        DEBUG_PUTS("[pms5003]: response timeout");
+    }
+    /* unlock mutex, always needs unlocking, either failed to acquire above
+       or succeeded and now needs unlocking */
+    mutex_unlock(&dev->internal.timeout);
+    return ret;
+}
+
+static void _cb_busy(void *arg)
+{
+    /* unlock busy mutex after elapsed time */
+    DEBUG_PUTS("[pms5003]: not busy");
+    mutex_unlock((mutex_t *)arg);
```

Comment:
```suggestion
    mutex_unlock(arg);
```

`(void *)` can be implicitly casted to any pointer type and explicit casts disable all of the little type safety the C has to offer.